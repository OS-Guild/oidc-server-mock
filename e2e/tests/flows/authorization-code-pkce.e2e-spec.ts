import * as crypto from 'crypto';

import * as dotenv from 'dotenv';
import { Browser, BrowserContext, chromium, Page } from 'playwright-chromium';

import clients from '../../config/clients-configuration.json';
import users from '../../config/user-configuration.json';
import { authorizationEndpoint, introspectEndpoint, tokenEndpoint, userInfoEndpoint } from '../../helpers';
import type { Client, User } from '../../types';

const testCases: User[] = users
  .map(u => ({
    ...u,
    toString: function () {
      return this.SubjectId;
    },
  }))
  .sort((u1, u2) => (u1.SubjectId < u2.SubjectId ? -1 : 1));

const base64URLEncode = (buffer: Buffer) =>
  buffer.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, '');

const sha256 = (buffer: crypto.BinaryLike) => crypto.createHash('sha256').update(buffer).digest();

describe('Authorization Code Flow (with PKCE)', () => {
  let codeVerifier: string;
  let code: string;
  let token: string;

  let browser: Browser;
  let context: BrowserContext;
  let page: Page;
  let client: Client;

  beforeAll(async () => {
    dotenv.config();

    browser = await chromium.launch();
    client = clients.find(c => c.ClientId === 'authorization-code-with-pkce-client-id');
    expect(client).toBeDefined();
  });

  beforeEach(async () => {
    context = await browser.newContext({ ignoreHTTPSErrors: true });
    page = await context.newPage();
  });

  afterEach(async () => {
    await page.close();
    await context.close();
  });

  afterAll(async () => {
    await browser.close();
  });

  describe.each(testCases)('- %s -', (user: User) => {
    test('Authorization Endpoint', async () => {
      codeVerifier = base64URLEncode(crypto.randomBytes(32));

      const codeChallenge = base64URLEncode(sha256(codeVerifier));

      const parameters = new URLSearchParams({
        client_id: client.ClientId,
        scope: client.AllowedScopes.join(' '),
        response_type: 'code',
        redirect_uri: client.RedirectUris?.[0].replace('*', 'www'),
        code_challenge: codeChallenge,
        code_challenge_method: 'S256',
        state: 'abc',
        nonce: 'xyz',
      });
      const redirectedUrl = await authorizationEndpoint(page, parameters, user, parameters.get('redirect_uri'));
      expect(redirectedUrl.searchParams.has('code')).toBeTruthy();
      code = redirectedUrl.searchParams.get('code');
    });

    test('Token Endpoint', async () => {
      const parameters = new URLSearchParams({
        client_id: client.ClientId,
        code,
        grant_type: 'authorization_code',
        redirect_uri: client.RedirectUris?.[0].replace('*', 'www'),
        code_verifier: codeVerifier,
        scope: client.AllowedScopes.join(' '),
      });

      token = await tokenEndpoint(parameters);
    });

    test('UserInfo Endpoint', async () => {
      await userInfoEndpoint(token);
    });

    test('Introspection Endpoint', async () => {
      await introspectEndpoint(token, 'some-app');
    });
  });
});
