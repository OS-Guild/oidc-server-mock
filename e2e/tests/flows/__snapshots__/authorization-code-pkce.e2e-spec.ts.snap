// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authorization Code Flow (with PKCE) - simple_user - Introspection Endpoint 1`] = `
{
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "amr": "pwd",
  "client_id": "authorization-code-with-pkce-client-id",
  "sub": "simple_user",
  "idp": "local",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Authorization Code Flow (with PKCE) - simple_user - Token Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "scope": [
    "openid",
    "profile",
    "email",
    "some-custom-identity",
    "some-app-scope-1"
  ],
  "amr": [
    "pwd"
  ],
  "client_id": "authorization-code-with-pkce-client-id",
  "sub": "simple_user",
  "idp": "local"
}
`;

exports[`Authorization Code Flow (with PKCE) - simple_user - UserInfo Endpoint 1`] = `
{
  "sub": "simple_user",
}
`;

exports[`Authorization Code Flow (with PKCE) - user_with_all_claim_types - Introspection Endpoint 1`] = `
{
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "amr": "pwd",
  "client_id": "authorization-code-with-pkce-client-id",
  "sub": "user_with_all_claim_types",
  "idp": "local",
  "some-app-user-custom-claim": "Oliver's Custom User Claim",
  "some-app-scope-1-custom-user-claim": "Oliver's Scope Custom User Claim",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Authorization Code Flow (with PKCE) - user_with_all_claim_types - Token Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "scope": [
    "openid",
    "profile",
    "email",
    "some-custom-identity",
    "some-app-scope-1"
  ],
  "amr": [
    "pwd"
  ],
  "client_id": "authorization-code-with-pkce-client-id",
  "sub": "user_with_all_claim_types",
  "idp": "local",
  "some-app-user-custom-claim": "Oliver's Custom User Claim",
  "some-app-scope-1-custom-user-claim": "Oliver's Scope Custom User Claim"
}
`;

exports[`Authorization Code Flow (with PKCE) - user_with_all_claim_types - UserInfo Endpoint 1`] = `
{
  "email": "oliver.hunter@gmail.com",
  "name": "Oliver Hunter",
  "some-custom-identity-user-claim": "Oliver's Custom User Claim",
  "sub": "user_with_all_claim_types",
}
`;

exports[`Authorization Code Flow (with PKCE) - user_with_custom_api_resource_claims - Introspection Endpoint 1`] = `
{
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "amr": "pwd",
  "client_id": "authorization-code-with-pkce-client-id",
  "sub": "user_with_custom_api_resource_claims",
  "idp": "local",
  "some-app-user-custom-claim": "Sam's Custom User Claim",
  "some-app-scope-1-custom-user-claim": "Sam's Scope Custom User Claim",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Authorization Code Flow (with PKCE) - user_with_custom_api_resource_claims - Token Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "scope": [
    "openid",
    "profile",
    "email",
    "some-custom-identity",
    "some-app-scope-1"
  ],
  "amr": [
    "pwd"
  ],
  "client_id": "authorization-code-with-pkce-client-id",
  "sub": "user_with_custom_api_resource_claims",
  "idp": "local",
  "some-app-user-custom-claim": "Sam's Custom User Claim",
  "some-app-scope-1-custom-user-claim": "Sam's Scope Custom User Claim"
}
`;

exports[`Authorization Code Flow (with PKCE) - user_with_custom_api_resource_claims - UserInfo Endpoint 1`] = `
{
  "email": "sam.tailor@gmail.com",
  "name": "Sam Tailor",
  "sub": "user_with_custom_api_resource_claims",
}
`;

exports[`Authorization Code Flow (with PKCE) - user_with_custom_identity_claims - Introspection Endpoint 1`] = `
{
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "amr": "pwd",
  "client_id": "authorization-code-with-pkce-client-id",
  "sub": "user_with_custom_identity_claims",
  "idp": "local",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Authorization Code Flow (with PKCE) - user_with_custom_identity_claims - Token Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "scope": [
    "openid",
    "profile",
    "email",
    "some-custom-identity",
    "some-app-scope-1"
  ],
  "amr": [
    "pwd"
  ],
  "client_id": "authorization-code-with-pkce-client-id",
  "sub": "user_with_custom_identity_claims",
  "idp": "local"
}
`;

exports[`Authorization Code Flow (with PKCE) - user_with_custom_identity_claims - UserInfo Endpoint 1`] = `
{
  "email": "jack.sparrow@gmail.com",
  "name": "Jack Sparrow",
  "some-custom-identity-user-claim": "Jack's Custom User Claim",
  "sub": "user_with_custom_identity_claims",
}
`;

exports[`Authorization Code Flow (with PKCE) - user_with_standard_claims - Introspection Endpoint 1`] = `
{
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "amr": "pwd",
  "client_id": "authorization-code-with-pkce-client-id",
  "sub": "user_with_standard_claims",
  "idp": "local",
  "active": true,
  "scope": "some-app-scope-1"
}
`;

exports[`Authorization Code Flow (with PKCE) - user_with_standard_claims - Token Endpoint 1`] = `
{
  "alg": "RS256",
  "typ": "JWT",
  "iss": "https://localhost:8443",
  "aud": "some-app",
  "scope": [
    "openid",
    "profile",
    "email",
    "some-custom-identity",
    "some-app-scope-1"
  ],
  "amr": [
    "pwd"
  ],
  "client_id": "authorization-code-with-pkce-client-id",
  "sub": "user_with_standard_claims",
  "idp": "local"
}
`;

exports[`Authorization Code Flow (with PKCE) - user_with_standard_claims - UserInfo Endpoint 1`] = `
{
  "email": "john.smith@gmail.com",
  "email_verified": "true",
  "name": "John Smith",
  "sub": "user_with_standard_claims",
}
`;
